### PR TITLE
feat: Integrate mistral

### LINKED ISSUE #742: [Feature Request] Integrate Mistral
### Required prerequisites

- [X] I have searched the [Issue Tracker](https://github.com/camel-ai/camel/issues) and [Discussions](https://github.com/camel-ai/camel/discussions) that this hasn't already been reported. (+1 or comment there if it has.)
- [ ] Consider asking first in a [Discussion](https://github.com/camel-ai/camel/discussions/new).

### Motivation

https://mistral.ai/news/mathstral/

### Solution

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Wendong-Fan**
## Description

Integrate Mistral AI natively including their embedding model, the current implementation doesn't support tool calling, will be further polished in next sprint. With small update for neo4j client setting

## Motivation and Context

`close #742 ` 

- [x] I have raised an issue to propose this change ([required](https://github.com/camel-ai/camel/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Implemented Tasks

- [ ] Subtask 1
- [ ] Subtask 2
- [ ] Subtask 3

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/camel-ai/camel/blob/master/CONTRIBUTING.md) guide. (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly. (*required for a bug fix or a new feature*)
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: camel/configs/mistral_config.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    """
+
+    temperature: Optional[float] = None
+    max_tokens: Optional[int] = None
+    top_p: Optional[float] = None
+    random_seed: Optional[int] = None
+    safe_mode: bool = False
+    safe_prompt: bool = False
+    response_format: Optional[Union[Dict[str, str], ResponseFormat]] = None
+    tools: Optional[list[OpenAIFunction]] = None
```
Comment (zechengz) [inline camel/configs/mistral_config.py:67]: Just curious whether all mistral models support tools? If not any way to check this so we may raise an error?

File: examples/models/mistral_model_example.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+from camel.agents import ChatAgent
+from camel.configs import MistralConfig
+from camel.messages import BaseMessage
+from camel.models import ModelFactory
+from camel.types import ModelPlatformType, ModelType
+
+model = ModelFactory.create(
+    model_platform=ModelPlatformType.MISTRAL,
+    model_type=ModelType.MISTRAL_LARGE,
+    model_config_dict=MistralConfig(temperature=0.2).__dict__,
```
Comment (zechengz) [inline examples/models/mistral_model_example.py:23]: nit: use temperature to 0 to make this example reproducible
```suggestion
    model_config_dict=MistralConfig(temperature=0).__dict__,
```

File: camel/configs/mistral_config.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    )
+
+    from camel.toolkits import OpenAIFunction
+
+
+@dataclass(frozen=True)
+class MistralConfig(BaseConfig):
+    r"""Defines the parameters for generating chat completions using the
+    Mistral API.
+
```
Comment (zechengz) [inline camel/configs/mistral_config.py:33]: Any reference url here?

File: examples/models/mistral_model_example.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+from camel.agents import ChatAgent
+from camel.configs import MistralConfig
+from camel.messages import BaseMessage
+from camel.models import ModelFactory
+from camel.types import ModelPlatformType, ModelType
+
+model = ModelFactory.create(
```
Comment (zechengz) [inline examples/models/mistral_model_example.py:20]: nit: maybe we can add something like `api_key = os.environ.get("MISTRAL_API_KEY")` and pass it to the model. Let user to know what env should be set more explicit / easier.

**Review by zechengz — APPROVED** [review]

File: camel/configs/mistral_config.py (hunk 3 above)
Comment (Wendong-Fan) [reply camel/configs/mistral_config.py:33]: reference updated

File: examples/models/mistral_model_example.py (hunk 4 above)
Comment (Wendong-Fan) [reply examples/models/mistral_model_example.py:20]: We are using `@api_keys_required("MISTRAL_API_KEY")` to do this, it's in the model file part

File: camel/configs/mistral_config.py (hunk 1 above)
Comment (Wendong-Fan) [reply camel/configs/mistral_config.py:67]: Currently, function calling is available for the following models:
Mistral Small
Mistral Large
Mixtral 8x22B
Mistral Nemo
refer: https://docs.mistral.ai/capabilities/function_calling/

I think we don't need to do the check from our side, if we add tools to the config for models don't support tool call we would get error from Mistral Client like 
```mistralai.exceptions.MistralAPIException: Status: 400. Message: {"object":"error","message":"Function calling is not enabled for this model","type":"invalid_request_error","param":null,"code":null}```

**Review by onemquan — APPROVED** [review]
LGTM

**Pull request merged**